text/1000-migrations.rst

```diff
@@ -332,47 +349,78 @@ The returned JSON conforms to the following pseudo-schema::
-----------------------------------
+The algorithm for obtaining the "proposed" data is as follows:
 
-The migrations system classifies the migrations operations into two categories:
-safe and unsafe, based on whether the operation is automatically reversible
-without losing any prior data.  For example, all ``CREATE`` operations are
-considered safe by definition, but also alterations to schema that doesn't
-involve data mutation, such as annotations, indexes, etc.  All other operations
-are classified as unsafe.
+1. Compute schema "A" from the state immediately preceding the
+   ``START MIGRATION`` command plus all statements in the "confirmed" list.
+
+2. Compute schema "B" from the SDL declaration passed to ``START MIGRATION``.
+
+3. Calculate the similarity matrix for each pair of objects of the same class
+   in schema "A" and schema "B" (e.g. compare every object type in schema "A"
+   with schema "B").  Only top-level objects are considered (i.e. object types,
+   abstract link definitions etc). The rows of the matrix represent objects
+   from "A", columns -- objects from "B", and the cells contain
+   *similarity index*, which is a floating point value from ``0.0`` to ``1.0``,
+   where ``0`` means "completely dissimilar" and ``1.0`` means "identical".
+   The similarity score is set to ``0`` if the command generated by this pair
+   in a previous run was rejected by ``ALTER CURRENT MIGRATION REJECT
+   PROPOSED``.
+
+4. Exclude all objects that are considered to be the same in both schemas from
+   the similarity matrix, i.e. exclude rows and columns that contain a ``1.0``
+   value.
+
+5. For each object in schema "B" in the similarity matrix, find the cell with
+   the greatest similarity score.  If the similarity score is greater than
+   ``0.6`` assume the object has been ``ALTER``-ed and generate an appropriate
+   DDL command.  Exclude the row and the column from further consideration.
+
+6. For each remaining row in the matrix, assume object deletion and generate
+   the appropariate ``DROP`` DDL.  For each remaining column in the matrix,
+   assume new object and generate the appropriate ``CREATE`` DDL.
+
+7. Sort the generated DDL commands according to the dependency topology.
+
+8. Pick the first DDL statement from the sorted list.  This is the new
+   "proposed" statement.
+
+
+ALTER CURRENT MIGRATION REJECT PROPOSED
+---------------------------------------
+
+This statement is only valid while inside the ``START MIGRATION`` block.
+When executed it marks the statement found in the "proposed" branch of
+``DESCRIBE CURRENT MIGRATION AS JSON`` as "unacceptable" and forces the
+server to recompute the diff with this command excluded.  This, effectively
+allows the client to implement the (y) / (n) flow for each proposed statement
+in the migration.  For (y) the client would send back the proposed statement
+to be recorded in the "confirmed" list, and for (n) an ``ALTER CURRENT
+MIGRATION REJECT PROPOSED`` is sent instead.
```

Comment:
FWIW this is potentially extendable to multiple-choice.

Reply:
I added a design discussion note about why multiple-choice in general is hard.